8332181: Deprecate for removal the java.net.MulticastSocket.setTTL/getTTL and the 2-arg send methods

jaikiran opened the pull request:
Can I please get a review of this change which proposes to deprecate for removal 3 methods on `java.net.MulticastSocket`? This addresses https://bugs.openjdk.org/browse/JDK-8332181.

As noted in that issue these methods have been deprecated since Java 1.2 and 1.4 days. They currently have replacement methods (noted in their javadoc) which have been in use for several releases. This commit updates these deprecated methods to deprecated for removal, to allow for their removal in a future release.

No new tests have been added and existing tests in tier1, tier2 and tier3 continue to pass.

jaikiran commented:
A CSR has been drafted for this change and is available for review here https://bugs.openjdk.org/browse/JDK-8332261.

robaho commented:
Part of what makes Java so great is that you can take a Java 1 program and run it under Java 22? with no modifications?

What does it cost to keep these methods around?

jaikiran commented:
Hello Robert,

> Part of what makes Java so great is that you can take a Java 1 program and run it under Java 22? with no modifications?
> 
> What does it cost to keep these methods around?

In this case, we do not want the applications to be using these methods. These methods accept a `ttl` value as `byte`. As noted in the javadoc of `setTTL()` method:

>The ttl is an unsigned 8-bit quantity, and so must be in the range {@code 0 <= ttl <= 0xFF }

The `byte` type doesn't allow for these values to be accomodated. That's why those methods were deprecated very long back in favour of the ones noted in their javadoc.

robaho commented:
That makes sense - thanks. I agree it is very, very old - just wanted more to understand the rationale as even though Java has deprecated a lot - I am not aware of any core removals - but only anecdotally.

dfuch commented on src/java.base/share/classes/java/net/NetMulticastSocket.java:
```diff
@@ -652,6 +652,7 @@ public Set<SocketOption<?>> supportedOptions() {
 
     @Deprecated
     @Override
+    @SuppressWarnings("removal")
```
Instead of adding `@SuppressWarning("removal")` here (and at other places below) - would it make sense to update the `@Deprecated` annotation and add `forRemoval = true`?

jaikiran replied:
We will still have to add the `@SuppressWarning("removal")` to these methods otherwise the compilation generates a warning which fails the build:

```
java.base/share/classes/java/net/NetMulticastSocket.java:655: warning: [removal] setTTL(byte) in MulticastSocket has been deprecated and marked for removal
    public void setTTL(byte ttl) throws IOException {
                ^
java.base/share/classes/java/net/NetMulticastSocket.java:673: warning: [removal] getTTL() in MulticastSocket has been deprecated and marked for removal
    public byte getTTL() throws IOException {
                ^
error: warnings found and -Werror specified
1 error
2 warnings

```
I went ahead and updated these internal `NetMulticastSocket` and `DatagramSocketAdaptor` classes to add the `forRemoval` attribute to the `@Deprecated` annotation to keep it consistent.

dfuch replied:
Ah - thank you. I had hoped that adding forRemoval = true to the annotation would allow us to get rid of the SuppressWarning.

irisclark reviewed (approved):
From the JCP POV, these changes look fine.  These changes match the associated CSR, which I've also Reviewed.

jaikiran commented:
Thank you Iris and Daniel for the reviews. I've moved the CSR to Finalized.

AlanBateman commented:
I think DatagramSocketImpl.getTTL/setTTL will need to be deprecated for removal at the same time.

Also I assume @Deprecated can be dropped from DatagramSocketAdaptor and NetMulticastSocket, they just override or call the deprecated methods so should only need `@SuppressWarnings("removal")`.

jaikiran commented:
Hello Alan,

> I think DatagramSocketImpl.getTTL/setTTL will need to be deprecated for removal at the same time.

I have now updated the PR to deprecate for removal these methods too. I have updated the CSR to note this change.

> Also I assume @deprecated can be dropped from DatagramSocketAdaptor and NetMulticastSocket, they just override or call the deprecated methods so should only need @SuppressWarnings("removal").

Done.

jaikiran commented:
Thank you everyone for the reviews. The CSR has been approved. I'll go ahead with the integration now.

jaikiran commented:
/integrate
